### ISSUE TITLE
Bug in scipy.signal.resample() when using a window

### ISSUE THREAD

**Issue by s9w** [scipy/scipy#6067]
The `signal.resample()` function accepts an optional parameter to specify a window in frequency domain. If that's done with a tuple with name and parameter, the window is retrieved from another function. Relevant code from [signaltools.py](https://github.com/scipy/scipy/blob/master/scipy/signal/signaltools.py#L1848):

```
if window is not None:
    if callable(window):
        # ...
    elif isinstance(window, ndarray):
        # ...
    else:
        W = fftpack.ifftshift(get_window(window, Nx))
```

However that `get_window()` function returns the window in time- and not frequency domain! Which is obviously not correct. The `get_window()` can be found [here in the windows.py](https://github.com/scipy/scipy/blob/master/scipy/signal/windows.py#L1610).

**Comment by Eric89GXL** [comment]
Looking at the docstring of `resample`, it says:
```
        Specifies the window applied to the signal in the Fourier
        domain.  See below for details.
```
The idea is that the window function, whatever it is, will be applied in the Fourier domain, after taking the FFT of the signal. Such a window could be a boxcar, hann window, etc. Often such windows are applied in the *time* domain (e.g., before doing an FFT), but they can also be applied in the *frequency* domain, which is where they will be used according to the docstring.

In this sense, I don't think there is a bug here, but maybe the docstring could be improved somehow. Does that make sense?

**Comment by s9w** [comment]
I still think this isn't right. `resample` is used on signals in time domain. The signal is the parameter `x`. It is then fourier transformed into `X` (note the capital letter). If a window parameter exists, we get to `X = X * W`, so it's multiplied by the window. So the window should be in fourier domain, all good so far.

The problem arises if you supply the window as a tuple. Then you get to the `get_window` part. And that function simply returns the time domain window (ie: not in fourier domain). Did that illustrate the issue more?

**Comment by Eric89GXL** [comment]
> And that [get_window] function simply returns the time domain window

Not quite -- I think this is the source of confusion. The `get_window` function just returns a bunch of discrete samples -- these could time domain samples or frequency domain samples, depending on the use case. The samples from `get_window` are not assumed to be time-domain samples here -- the samples are considered to be already in the Fourier domain.

The docstring says:
```
    The argument `window` controls a Fourier-domain window that tapers
    the Fourier spectrum before zero-padding to alleviate ringing in
    the resampled values for sampled signals you didn't intend to be
    interpreted as band-limited.
```
So the idea is that any set of discrete values passed in as the `window` parameter should already be in the Fourier domain. Thus when this comes up:
```
    For any other type of `window`, the function `scipy.signal.get_window`
    is called to generate the window.
```
The idea is that whatever `get_window` returns is considered to already be in the Fourier domain.

To achieve time-domain windowing like I think you want, you'd simply apply the time-domain window before passing to `resample`. Make sense?

**Comment by s9w** [comment]
> The samples from get_window are not assumed to be time-domain samples here -- the samples are considered to be already in the Fourier domain.

But they aren't. They are in time domain.

> The idea is that whatever get_window returns is considered to already be in the Fourier domain.

Again: They aren't. They are in time domain.

Let's look at at the example (and documentation) from the kaiser sample. Simply because it's what I use and is the default. Source is [here](https://github.com/scipy/scipy/blob/master/scipy/signal/windows.py#L948).

![xx](http://i.imgur.com/bALhdco.png)

the get_window returns the top function, which is in time domain. What we want in the `resample` is the bottom part, which is the fft'd thing.

Now you can make a case that the actual fault is in the get_window function because it returns time-domain windows. But still, currently this is not correct if I'm not mistaken.

**Comment by Eric89GXL** [comment]
> They are in time domain.

What I'm trying to say is that this is an assumption, and it's not always true. That function gives samples -- whether they are in the time domain or frequency domain depends on the application where the window will be used.

> Let's look at at the example (and documentation) from the kaiser sample.

Yes, in this example, the samples from `get_window` are taken to be time-domain samples, as they are applied to a signal in the time domain.

> Now you can make a case that the actual fault is in the get_window function because it returns time-domain windows.

It doesn't return time-domain windows -- it returns samples (in some domain, up to the user!) of the given window. The samples from those windows could be in the time domain, or the frequency domain, or some other domain. Look at the docstring for `get_window` -- it doesn't specify what domain the samples are in. The examples happen to use them in the time domain, but that's only one such application. There is nothing to stop you from applying a `kaiser` window to the Fourier components of a signal, e.g. to accomplish a lowpass. To wit:
```
import numpy as np
import matplotlib.pyplot as plt
from scipy import fftpack
from scipy.signal import get_window
x = np.random.RandomState(0).randn(100)
x_lp = fftpack.ifft(fftpack.fft(x) * fftpack.ifftshift(get_window(('kaiser', 5), 100))).real
x_tw = x * get_window(('kaiser', 5), 100)
plt.plot(x)
plt.plot(x_lp)
plt.plot(x_tw)
```
Resulting in (blue original, green the low-passed `x_lp`, red the time-domain-windowed `x_tw`):

![figure_1](https://cloud.githubusercontent.com/assets/2365790/14537169/e12cdeae-0243-11e6-893f-13ab8247941c.png)

Here `x_lp` is processed in the way that the docstring of `resample` implies -- namely that the window is assumed to be a window already in the frequency domain. You can see from this example how `get_window` samples can be taken to be in either the frequency domain (as in how they're used in `x_lp` and `resample) or in the time domain (as they're used to create `x_tw`).

The docstring for `resample` says:
```
    The argument `window` controls a Fourier-domain window that tapers
    the Fourier spectrum before zero-padding to alleviate ringing in
    the resampled values for sampled signals you didn't intend to be
    interpreted as band-limited.
```
The window tapers the Fourier spectrum, not the time-domain (input) signal. The window function is applied in the frequency domain to change the band-limiting during the transform, which is different from the application in the time domain that you're thinking of, which admittedly is a more common way to think of windowing.

**Comment by s9w** [comment]
>>    Let's look at at the example (and documentation) from the kaiser sample.

>Yes, in this example, the samples from get_window are taken to be time-domain samples, as they are applied to a signal in the time domain.

They aren't applied to any signal. The window itself gets drawn, as well as its fourier transform. (In the kaiser window example, I mean)

---

Your example uses random numbers. Since the core of this problem is the behavior of signals and filters in time and frequency domains, I think that kind of "signal" makes it very hard to get an intuitive understanding of what's going on. So I've tried to illustrate with a little more meaningful example and in some more detail in this notebook: https://gist.github.com/s9w/438df149e81166701aab7ea8ce820ade

---

The heart of the problem really is two part:

1) `resample()` applies the window in frequency domain, so it should be in frequency domain.
2) `get_window()` returns a time-domain window, at least for the case of kaiser.

And the problem being that

3) This is not correct

I think we agree on 2)? For 1), I'll throw in this commentated `resample()` code:

```python
def resample(x, num, t=None, axis=0, window=None):
    x = asarray(x) # x is in time domain
    X = fftpack.fft(x, axis=axis) # X is fft(x), so X is in frequency (fourier) domain
    Nx = x.shape[axis]
    if window is not None:
        if callable(window):
            # ...
        elif isinstance(window, ndarray):
            # ...
        else:
            W = fftpack.ifftshift(get_window(window, Nx))
            # get_window() returns kaiser in time domain. That get's ifftshift'ed, which only makes sense if it would be in frequncy domain
        newshape = [1] * x.ndim
        newshape[axis] = len(W)
        W.shape = newshape
X = X * W # and finally, the frequency-domain X is multiplied by the time-domain W
```

For 3) I'm not sure.

**Comment by Eric89GXL** [comment]
> Your example uses random numbers. Since the core of this problem is the behavior of signals and filters in time and frequency domains, I think that kind of "signal" makes it very hard to get an intuitive understanding of what's going on.

Such white noise is so named because the spectrum is ideally white, i.e. it contains all frequencies. As such it's usually a useful tool for examples -- at least that's the motivation for using it. Hopefully you could at least see from my example that applying the window as a time-domain window is very different from applying the window as a frequency-domain window, and intuitively understand why (as a time-domain tapering or freq-domain lowpass).

I think the problem isn't as you say, but rather with your *assumption* that:

> get_window() returns a time-domain window, at least for the case of kaiser.

This isn't true. It returns a window. Whether that window is in the time domain or frequency domain is up to the particular use case.

Consider a boxcar window. Is that always a time-domain window? A boxcar window can be applied in the time domain to achieve e.g. cutting off a signal, or in the frequency domain to achieve e.g. a lowpass. Both are valid use cases. The latter is what `resample` would do.

The kaiser window is the same as the boxcar window in this respect. Both just return discrete samples. These could be in the frequency domain or the time domain *already*, depending on the application you want it for. You can of course consider one to be in the time domain and look at its frequency response (as you keep doing), but you could just as easily (and validly) assume it's a frequency-domain window already (as resample does), and take the IFFT to look at its time-domain function.

Perhaps I'm not explaining it that well -- maybe @endolith or @e-q could do a better job...

**Comment by e-q** [comment]
I think I can see where both of you are coming from. My wall-of-opinion follows...

TL;DR:
- I also find the treatment of the `window` argument surprising, but mostly due to its name
- I don't think `resample` is doing anything categorically **wrong**
- `resample` does lack proper time-domain windowing, but to fight spectral leakage, not to make the signal smooth

- - -

I'll start by saying, I am also surprised by the treatment of the `window` argument in `resample`, however I don't think its calculations are necessarily _wrong_. If I were to write this kind of function today, I would call what is currently `window` something like `filter` or `fft_weights` (and let it be complex), and have the `window` argument be an application of a window in the time domain, as is the case in all of the `signal.spectral` functions. If I didn't read the docstring carefully, this is what I would've assumed the `window` argument controls.

The kind of operation that is happening now with `window` is not uncommon, I've seen it used for low/high/bandpass filtering / frequency selectivity. It is a little surprising to me to use the assortment of window functions to this end; I would be more inclined to use just a step function-like selection of frequencies to include, or something like a power-law rolloff, or maybe some complex transfer function. While it is weird to me to be using window functions for this purpose, this doesn't mean that anything about it is wrong; the docstring is upfront about what is happening. (And @Eric89GXL is right, the window functions are just _functions_, their application in whatever domain is up to us)

How the window is applied in `window_trouble` notebook doesn't strike me as entirely appropriate, either. Normally, time-domain windowing is applied via multiplication in the time-domain, which leads to _convolution_ in the frequency domain; this is how it controls spectral leakage between bins. In the notebook, the FFT is still just getting multiplied by a some array of numbers that weight the FFT bins differently; you're still multiplying in the frequency domain, leading to convolution in the time domain - a common method of signal smoothing. This is a different effect than what time-domain windowing is generally intended to achieve.

That is: I don't agree that the signal becoming smooth is the desired behavior from `resample`; I may want to only filter out components that are out of the new band in order to prevent aliasing, or I may want to only extract certain frequencies of interest. The current implementation of `window` lets me do those things, but convolving the time-domain signal with window functions is always a low-pass. However, `resample` does not have any ability to minimize spectral leakage via time-domain windowing, which is a real deficiency.

**Comment by Eric89GXL** [comment]
> However, resample does not have any ability to minimize spectral leakage via time-domain windowing, which is a real deficiency.

I don't see this as a deficiency so much as a reasonable design choice. You can accomplish such time-domain windowing to deal with edge artifacts or spectral leakage with `resample(x * window, ...)` in your own code, whereas the Fourier-domain windowing occurs in the middle of transforms (after FFT, before truncation and IFFT) so it has to be a parameter...

**Comment by endolith** [comment]
I don't understand the purpose of the frequency-domain window.  I was thinking it was like the FIR antialiasing filter in [`decimate`](http://scipy.github.io/devdocs/generated/scipy.signal.decimate.html), but it doesn't seem to be?

[Matlab's only allows for Kaiser windows](http://www.mathworks.com/help/signal/ref/resample.html#bungy76)?  But I still don't understand what the benefit is supposed to be.

But anyway I agree that time-domain windowing is something you do outside of the function and is not related to the `window` parameter.

**Comment by Eric89GXL** [comment]
MATLAB's `resample` is like our `resample_poly` in that it is a time domain
method using polyphase filtering -- the window they refer to is used as a
time-domain FIR antialiasing filter just like in decimate. Our `resample`
uses frequency domain methods, so its use of a frequency domain window is
unfortunately not exactly analogous (even though you can use the same
window in one or the other).

You could however think of the frequency domain multiplication by a window
function as equivalent to the (circular) convolution in time with its IFFT,
so in that sense the freq domain window does act like an antialiasing
filter, even though the truncation of the FFT (or extension with extra
zeros) will already perform antialiasing for you. In other words, the
window gives you some control over how that antialiasing filter is applied
(brick wall or more gradual rolloff).

**Comment by endolith** [comment]
Hmmm: 
![resample with white noise](https://cloud.githubusercontent.com/assets/58611/14578230/de7a48e6-0350-11e6-8175-e3bff7335079.png)

**Comment by Eric89GXL** [comment]
That's what I'd expect from resample called (with a non-boxcar argument)
based on the docstring...?

**Comment by Eric89GXL** [comment]
The extent to which it is actually useful or used in practice I'm not sure
about, but I'm pretty sure the docstring and code are consistent about what
is happening, and there is nothing inherently wrong with it... Maybe we
should git blame back to the original author to see what they had in mind.

**Comment by e-q** [comment]
Closing since there doesn't seem to be anything actionable here at this point. Specific improvements can be proposed in a new PR.

**Comment by endolith** [comment]
I haven't read through in detail, but if resample is using the window functions to generate an FIR low pass filter, shouldn't it multiply by the FFT of the window in the frequency domain, not the window itself?

**Comment by Eric89GXL** [comment]
That is what this discussion is about, so you'll have to read through in
detail. I think this is a documentation problem
